src/py21cmfast/src/ps.c

```diff
@@ -2217,27 +2252,125 @@ float GaussLegendreQuad_Nion_MINI(int Type, int n, float growthf, float M2, floa
         .LimitMass_Fesc = Mlim_Fesc_MINI
     };
 
-    if(delta2 > delta1){
-        return 1.;
+
+
+
+    if(delta2 > delta1*0.9999) {
+        result = 1.;
+        return result;
+    }
+
+
+    if(FAST_FCOLL_TABLES){ //JBM: Fast tables. Assume sharp Mturn, not exponential cutoff.
+
+
+      if(MassTurnover_upper <= MassTurnover){
+        return 1e-40; //in sharp cut it's zero
+      }
+
+      double LogMass=log(MassTurnover);
+      int MassBin = (int)floor( (LogMass - MinMass )*inv_mass_bin_width );
+      double MassBinLow = MinMass + mass_bin_width*(double)MassBin;
+      double sigmaM1 = Sigma_InterpTable[MassBin] + ( LogMass - MassBinLow )*( Sigma_InterpTable[MassBin+1] - Sigma_InterpTable[MassBin] )*inv_mass_bin_width;
+      nu_lower_limit = (delta1-delta2)*(delta1-delta2)/(sigmaM1*sigmaM1-sigma2*sigma2)/growthf/growthf;
+
+      LogMass = log(MassTurnover_upper);
+      MassBin = (int)floor( (LogMass - MinMass )*inv_mass_bin_width );
+      MassBinLow = MinMass + mass_bin_width*(double)MassBin;
+      double sigmaM2 = Sigma_InterpTable[MassBin] + ( LogMass - MassBinLow )*( Sigma_InterpTable[MassBin+1] - Sigma_InterpTable[MassBin] )*inv_mass_bin_width;
+      nu_higher_limit = (delta1-delta2)*(delta1-delta2)/(sigmaM2*sigmaM2-sigma2*sigma2)/growthf/growthf;
+
+
+      //note we keep nupivot1 just in case very negative delta makes it reach that nu
+      LogMass = log(1.5e9); //jbm could be done outside and it'd be even faster
```

Comment:
This (along with the other one below) should definitely be defined in Globals.h (or Constants.h). It is something that is unlikely to ever change, but maybe it might under certain conditions so should probably be in Globals.h

Reply:
I'm not sure it should be a global variable, since it's not something likely to change (since the sigma(M) fit is good to ~10%, which is worse than the errors in the cosmological parameters). I have moved it to constants.h, which is already better than having numbers in the code.